core/eolearn/tests/test_utils/test_testing.py

```diff
@@ -0,0 +1,140 @@
+import datetime as dt
+from dataclasses import dataclass
+from typing import Any, Dict
+
+import numpy as np
+import pytest
+
+from sentinelhub import CRS, BBox
+from sentinelhub.testing_utils import assert_statistics_match
+
+from eolearn.core.constants import FeatureType
+from eolearn.core.types import FeatureSpec, FeaturesSpecification
+from eolearn.core.utils.parsing import parse_features
+from eolearn.core.utils.testing import PatchGeneratorConfig, generate_eopatch
+
+
+def test_generate_eopatch_set_bbox_timestamps() -> None:
+    bbox = BBox((0, 0, 10, 10), crs=CRS("EPSG:32633"))
+    timestamps = [dt.datetime(2019, 1, 1)]
+    patch = generate_eopatch((FeatureType.DATA, "bands"), bbox=bbox, timestamps=timestamps)
+
+    assert patch.bbox == bbox
+    assert patch.timestamps == timestamps
+
+    assert patch[(FeatureType.DATA, "bands")].shape[0] == len(timestamps)
+
+
+@pytest.mark.parametrize(
+    "feature",
+    [
+        (FeatureType.META_INFO, "meta_info"),
+        (FeatureType.VECTOR, "vector"),
+        (FeatureType.VECTOR_TIMELESS, "vector_timeless"),
+    ],
+)
+def test_generate_eopatch_fails(feature: FeatureSpec) -> None:
+    with pytest.raises(ValueError):
+        generate_eopatch(feature)
+
+
+@pytest.mark.parametrize("seed", [0, 1, 42, 100])
+@pytest.mark.parametrize(
+    "features",
+    [
+        {},
+        {FeatureType.DATA: ["bands, CLP"]},
+        {FeatureType.DATA: ["bands, CLP"], FeatureType.MASK_TIMELESS: "LULC"},
+        {
+            FeatureType.DATA: "data",
+            FeatureType.MASK: "mask",
+            FeatureType.SCALAR: "scalar",
+            FeatureType.LABEL: "label",
+            FeatureType.DATA_TIMELESS: "data_timeless",
+            FeatureType.MASK_TIMELESS: "mask_timeless",
+            FeatureType.SCALAR_TIMELESS: "scalar_timeless",
+            FeatureType.LABEL_TIMELESS: "label_timeless",
+        },
+    ],
+)
+def test_generate_eopatch_seed(seed: int, features: FeaturesSpecification) -> None:
+    patch1 = generate_eopatch(features, seed=seed)
+    patch2 = generate_eopatch(features, seed=seed)
+    assert patch1 == patch2
+
+
+@dataclass
+class GenerateTestCase:
+    data_features: FeaturesSpecification
+    seed: int
+    expected_statistics: Dict[FeatureSpec, Dict[str, Any]]
+
+
+@pytest.mark.parametrize(
+    "test_case",
+    [
+        GenerateTestCase(
+            seed=3,
+            data_features=(FeatureType.MASK, "CLM"),
+            expected_statistics={
+                (FeatureType.MASK, "CLM"): {
+                    "exp_shape": (5, 98, 151, 2),
+                    "exp_min": 0,
+                    "exp_max": 255,
+                    "exp_mean": 127.38907960535208,
+                    "rel_delta": 0.0001,
```

Comment:
since `rel_delta` doesn't really get changed, maybe hardcode it in the test (less clutter in test cases)

I would also keep only a few values after the decimal point, also so that there is less text to look at. Otherwise it's fine

Reply:
done